eth/beacon/state_machines/base.py

```diff
@@ -0,0 +1,95 @@
+from abc import (
+    ABC,
+)
+from typing import (
+    Tuple,
+    Type,
+)
+
+from eth.utils.datatypes import (
+    Configurable,
+)
+
+from eth.beacon.db.chain import BaseBeaconChainDB
+from eth.beacon.types.blocks import BaseBeaconBlock
+from eth.beacon.types.states import BeaconState
+
+
+from .state_transitions import (
+    BaseStateTransition,
+)
+from .configs import (  # noqa: F401
+    BeaconConfig,
+)
+
+
+class BaseBeaconStateMachine(Configurable, ABC):
+    fork = None  # type: str
+    chaindb = None  # type: BaseBeaconChainDB
+    config = None  # type: BeaconConfig
+
+    block = None  # type: BaseBeaconBlock
+    state = None  # type: BeaconState
+
+    block_class = None  # type: Type[BaseBeaconBlock]
+    state_class = None  # type: Type[BeaconState]
+    state_transition_class = None  # type: Type[BaseStateTransition]
+
+
+class BeaconStateMachine(BaseBeaconStateMachine):
+    def __init__(self,
+                 chaindb: BaseBeaconChainDB,
+                 block: BaseBeaconBlock,
+                 state: BeaconState) -> None:
+        # TODO: get state from DB, now it's just a stub!
+        self.state = state
+
+    @classmethod
+    def get_block_class(cls) -> Type[BaseBeaconBlock]:
```

Comment:
@hwwhww @djrtwo I think we should put all these methods as `abstract` methods on the `BaseBeaconStateMachine` just to have a formal interace without implementation. Similar to what I stubbed in https://github.com/ethereum/py-evm/issues/1616#issuecomment-448297218

Reply:
@cburgdorf 
👍 on having @abstract methods on `BaseBeaconStateMachine`!

I suppose we still want to have `class StateMachine(BaseStateMachine)` (copied from `class VM(BaseVM)`) ? And  then define `class SerenityStateMachine(StateMachine)`? <-- different from https://github.com/ethereum/py-evm/issues/1616#issuecomment-448297218

Reply:
Yes, the `StateMachine(BaseStateMachine)` can implement everything that is generic enough to be implemented on that level. :+1: